LEAF_Request_Portal/FormWorkflow.php

```diff
@@ -61,177 +63,206 @@ public function isActive(): bool
+            $this->cache["dependencyID{$depID}"] = $this->db->prepared_query($strSQL, $vars);
         }
+        
+        return $this->cache["dependencyID{$depID}"];
+    }
 
-        $steps = array();
-        $vars = array(':recordID' => $this->recordID);
-        $strSQL = 'SELECT dependencyID, recordID, stepID, stepTitle, blockingStepID, workflowID, serviceID, filled, stepBgColor, stepFontColor, stepBorder, description, indicatorID_for_assigned_empUID, indicatorID_for_assigned_groupID, jsSrc, userID, requiresDigitalSignature FROM records_workflow_state
+    private function getServiceQuadrads(int $serviceID) {
+        if(!isset($this->cache["serviceQuadrads{$serviceID}"])) {
+            $quadGroupIDs = $this->login->getQuadradGroupID(); // get csv of ints
+            $vars3 = array(
+                ':serviceID' => $serviceID,
+            );
+            $strSQL = 'SELECT * FROM services
+                WHERE groupID IN ('.$quadGroupIDs.')
+                AND serviceID = :serviceID';
+            $this->cache["serviceQuadrads{$serviceID}"] = $this->db->prepared_query($strSQL, $vars3);
+        }
+
+        return $this->cache["serviceQuadrads{$serviceID}"];
+    }
+
+    public function getActionable(object $form, array $records): array {
+        $numRecords = count($records);
+        if ($numRecords == 0) {
+            return $records;
+        }
+
+        $recordIDs = '';
+
+        foreach ($records as $item) {
+            $recordIDs .= (int)$item['recordID'] . ',';
+        }
+        $recordIDs = trim($recordIDs, ',');
+        
+        $strSQL = "SELECT dependencyID, recordID, stepID, stepTitle, blockingStepID, workflowID, serviceID, filled, stepBgColor, stepFontColor, stepBorder, description, indicatorID_for_assigned_empUID, indicatorID_for_assigned_groupID, jsSrc, userID, requiresDigitalSignature FROM records_workflow_state
             LEFT JOIN records USING (recordID)
             LEFT JOIN workflow_steps USING (stepID)
             LEFT JOIN step_dependencies USING (stepID)
             LEFT JOIN dependencies USING (dependencyID)
             LEFT JOIN records_dependencies USING (recordID, dependencyID)
-            WHERE recordID = :recordID';
-        $res = $this->db->prepared_query($strSQL, $vars);
+            WHERE recordID IN ({$recordIDs})";
+        $res = $this->db->prepared_query($strSQL, []);
 
-        $numRes = count($res);
-        if ($numRes > 0)
-        {
-            if ($numRes == 1 && $res[0]['filled'] == 1) {
-                $res[0]['filled'] = 0;
-            }
+        foreach ($res as $i => $record) {
+            // override access if user is in the admin group
+            $res[$i]['isActionable'] = $this->login->checkGroup(1); // initialize isActionable
 
-            for ($i = 0; $i < $numRes; $i++)
+            // check permissions
+            $res2 = $this->getDependency($res[$i]['dependencyID']);
+
```

Comment:
Could just check if it is actionable at the top and continue if true.

Reply:
Definitely more efficient